src/main/java/subway/domain/Line.java

```diff
@@ -29,20 +63,150 @@ public String getName() {
+                .filter(station -> station.equals(upStation))
+                .findFirst();
+        Optional<Station> downStationOptional = getStations().stream()
+                .filter(station -> station.equals(downStation))
+                .findFirst();
+
+        if (upStationOptional.isPresent()) {
+            // 기존 역이 upStation이다
+            Optional<Edge> edgeOptional = edges.stream()
+                    .filter(edge -> edge.getUpStation().equals(upStationOptional.get()))
+                    .findFirst();
+
+            if (edgeOptional.isPresent()) {
+                // 거리비교하면 됨
+                Edge targetEdge = edgeOptional.get();
+                if (distance < targetEdge.getDistance()) { // 성공적으로 추가 가능
+                    Edge newEdge1 = new Edge(upStation, downStation, distance);
+                    Edge newEdge2 = new Edge(downStation, targetEdge.getDownStation(),
+                            targetEdge.getDistance() - distance);
+                    int removedIndex = edges.indexOf(targetEdge);
+                    edges.remove(removedIndex);
+                    edges.add(removedIndex, newEdge2);
+                    edges.add(removedIndex, newEdge1);
+                } else {
+                    throw new IllegalArgumentException("추가하려는 거리가 기존의 거리보다 깁니다.");
+                }
+
+            } else {
+                // 종점이니까 그냥 추가하면 돼
+                edges.add(new Edge(upStation, downStation, distance));
+            }
+        } else {
+            // 기존 역이 downStation이다.
+            Optional<Edge> edgeOptional = edges.stream()
+                    .filter(edge -> edge.getDownStation().equals(downStationOptional.get()))
+                    .findFirst();
+
+            if (edgeOptional.isPresent()) {
+                // 거리비교하면 됨
+                Edge targetEdge = edgeOptional.get();
+                if (distance < targetEdge.getDistance()) {
+                    Edge edge1 = new Edge(targetEdge.getUpStation(), upStation, targetEdge.getDistance() - distance);
+                    Edge edge2 = new Edge(upStation, downStation, distance);
+                    int removedIndex = edges.indexOf(targetEdge);
+                    edges.remove(removedIndex);
+                    edges.add(removedIndex, edge2);
+                    edges.add(removedIndex, edge1);
+                } else {
+                    throw new IllegalArgumentException("추가하려는 거리가 기존의 거리보다 깁니다.");
+                }
+            } else {
+                // 종점이니까 그냥 추가하면 돼
+                edges.add(0, new Edge(upStation, downStation, distance));
+            }
+        }
+    }
+
+    private boolean isNothingExist(Station upStation, Station downStation) {
+        return !getStations().contains(upStation) && !getStations().contains(downStation);
+    }
+
+    private boolean isAlreadyExistBoth(Station upStation, Station downStation) {
+        return getStations().containsAll(List.of(upStation, downStation));
```

Comment:
```suggestion
        return new HashSet<>(getStations()).containsAll(List.of(upStation, downStation));
```
성능상 위와 같이 작성 하는게 더 유리하다고 합니다 :)

Reply:
1. getStations().containsAll(List.of(upStation, downStation))
- 리스트에서 containsAll 메서드를 호출하여 해당 리스트가 모든 요소를 포함하는지 확인
- 리스트에서 항목을 찾을 때는 요소들을 처음부터 끝까지 순차적으로 탐색 -> 리스트의 크기가 커지면 검색 속도가 선형적으로 증가 -> 시간 복잡도 O(N)
- 리스트의 크기가 커질수록 선형적인 시간 복잡도를 가지게 되어 검색 속도가 느려질 가능성이 있다.

2. new HashSet<>(getStations()).containsAll(List.of(upStation, downStation))
- 먼저 기존 리스트를 HashSet으로 변환한 다음, containsAll 메서드를 사용하여 모든 요소를 포함하는지 확인
- HashSet은 내부적으로 해시 테이블을 사용하며, 각 요소를 고유한 버킷에 할당
- HashSet에서는 요소를 검색할 때 해시 함수를 사용하여 요소가 저장된 버킷을 직접 찾아서 검색 -> 이 때, 해시 함수를 사용하기 때문에 검색 시간은 요소의 개수와는 무관하게 상수 시간에 근접 -> 시간 복잡도 O(1)
- 따라서 큰 리스트의 경우 HashSet을 사용하면 검색 속도를 개선할 수 있다.

=> 리스트의 크기가 크거나 항목의 검색이 빈번하게 발생하는 경우, HashSet을 사용하여 검색 속도를 개선할 수 있다. 그러나 리스트의 크기가 작거나 항목 검색이 드물게 발생하는 경우, 두 코드 간의 성능 차이는 미미할 수 있다.

Reply:
새로운 정보를 알게되었네요! hashset을 통해 검색하도록 수정했습니다!